cilium/socket_option_source_address.h

```diff
@@ -0,0 +1,146 @@
+      : source_identity_(source_identity),
+        original_source_address_(std::move(original_source_address)),
+        ipv4_source_address_(std::move(ipv4_source_address)),
+        ipv6_source_address_(std::move(ipv6_source_address)) {
+    ENVOY_LOG(debug,
+              "Cilium SourceAddressSocketOption(): source_identity: {}, source_addresses: {}/{}/{}",
+              source_identity, original_source_address_ ? original_source_address_->asString() : "",
+              ipv4_source_address_ ? ipv4_source_address_->asString() : "",
+              ipv6_source_address_ ? ipv6_source_address_->asString() : "");
+  }
+
+  absl::optional<Network::Socket::Option::Details>
+  getOptionDetails(const Network::Socket&,
+                   envoy::config::core::v3::SocketOption::SocketState) const override {
+    return absl::nullopt;
+  }
+
+  bool setOption(Network::Socket& socket,
+                 envoy::config::core::v3::SocketOption::SocketState state) const override {
+    // Only set the option once per socket
+    if (state != envoy::config::core::v3::SocketOption::STATE_PREBIND) {
+      ENVOY_LOG(trace, "Skipping setting socket ({}) source address, state != STATE_PREBIND",
+                socket.ioHandle().fdDoNotUse());
+      return true;
+    }
+
+    auto ipVersion = socket.ipVersion();
+    if (!ipVersion.has_value()) {
+      ENVOY_LOG(critical, "Socket address family is not available, can not choose source address");
+      return false;
+    }
+
+    Network::Address::InstanceConstSharedPtr source_address = original_source_address_;
+    if (!source_address && (ipv4_source_address_ || ipv6_source_address_)) {
+      // Select source address based on the socket address family
+      source_address = ipv6_source_address_;
+      if (*ipVersion == Network::Address::IpVersion::v4) {
+        source_address = ipv4_source_address_;
+      }
+    }
+
+    if (!source_address) {
+      ENVOY_LOG(trace, "Skipping restore of local address on socket: {} - no source address",
+                socket.ioHandle().fdDoNotUse());
+      return true;
+    }
+
+    socket.connectionInfoProvider().setLocalAddress(std::move(source_address));
+    ENVOY_LOG(trace, "Successfully restored local address on socket: {}",
+              socket.ioHandle().fdDoNotUse());
+
+    return true;
+  }
+
+  template <typename T> void addressIntoVector(std::vector<uint8_t>& vec, const T& address) const {
+    const uint8_t* byte_array = reinterpret_cast<const uint8_t*>(&address);
+    vec.insert(vec.end(), byte_array, byte_array + sizeof(T));
+  }
+
+  void hashKey(std::vector<uint8_t>& key) const override {
+    // Source address is more specific than policy ID. If using an original
+    // source address, we do not need to also add the source security ID to the
+    // hash key. Note that since the identity is 3 bytes it will not collide
+    // with neither an IPv4 nor IPv6 address.
+    if (original_source_address_) {
```

Comment:
Currently, the `hashKey` implementation only considers the original source address - but not the configurable ipv4/ipv6 source addresses that are passed in most of the cases.

I wonder whether this is correct - or whether this was missed when the ingress source addresses have been introduced. maybe we should use the source addresses for building the hash in these cases too (instead of falling back to use the source identity.

Reply:
Could be an issue, but when L7LB is used (E/W or N/S), then the original source address is not kept: https://github.com/cilium/proxy/blob/main/cilium/bpf_metadata.cc#L427

So the connection pool (hashKey) is restricted on the source identity only, which likely is WORLD etc.